### PR TITLE
Add possibility to copy ConfigFile to host

### PR DESCRIPTION
**Pull request by afbjorklund**
* Refactor PortForward guest/host processing
* Add possibility to copy ConfigFile to host

Issue #1300

### CODE REVIEW

File: examples/k3s.yaml [hunk 1]
```diff
@@ -54,11 +53,12 @@ probes:
     The k3s kubeconfig file has not yet been created.
     Run "limactl shell k3s sudo journalctl -u k3s" to check the log.
     If that is still empty, check the bottom of the log at "/var/log/cloud-init-output.log".
+configFiles:
```
Comment (AkihiroSuda) [inline examples/k3s.yaml:56]:
```suggestion
copyToHost:
```

File: examples/default.yaml [hunk 2]
```diff
@@ -336,6 +336,16 @@ networks:
...
 # # Any port still not matched by a rule will not be forwarded (ignored)
 
+# Config files. Copied after provisioning scripts have been completed.
+# copyToHost:
+# - guest: "/etc/myconfig.cfg"
+#   host: myconfig
+# # default: mode: system
+# # "guest" can include these template variables: {{.Home}}, {{.UID}}, and {{.User}}.
+# # "host" can include {{.Home}}, {{.Dir}}, {{.Name}}, {{.UID}}, and {{.User}}.
+# # "mode" can be "system" (with root) or "user" (without root), similar to provision
```
[the remark this answers is not in the document]
Comment (afbjorklund) [reply examples/default.yaml:346]: Just seemed wrong to hardcode sudo/root, for those cases where it is not needed.

File: pkg/limayaml/defaults.go [hunk 3]
```diff
@@ -527,6 +532,47 @@ func FillDefault(y, d, o *LimaYAML, filePath string) {
 	y.CACertificates.Certs = caCerts
 }
 
+func processGuest(format string) (bytes.Buffer, error) {
```
Comment (jandubois) [inline pkg/limayaml/defaults.go:535]: I find `process` kind of meaningless, as it doesn't really tell me what is happening.

Maybe `expandGuestTemplate` (and `expandHostTemplate`) are better names?

**Review by jandubois — COMMENTED** [review]
Except for the file permissions this seems to be ready for merging, I think, so maybe remove the Draft status?

File: pkg/limayaml/defaults.go (hunk 3 above)
Comment (afbjorklund) [reply pkg/limayaml/defaults.go:535]: Those names sound better, I was struggling a bit

Comment (afbjorklund) [reply pkg/limayaml/defaults.go:535]: Or even executeGuestTemplate, like template.Execute

**Review by jandubois — APPROVED** [review]
Thanks, LGTM.

@AkihiroSuda Please merge if you have no objections!

File: pkg/store/filenames/filenames.go [hunk 4]
```diff
@@ -45,8 +45,11 @@ const (
 	VzIdentifier       = "vz-identifier"
 	VzEfi              = "vz-efi"
 
-	// SocketDir is the default location for forwarded sockets with a relative paths in HostSocket
-	SocketDir = "sock"
+	// InstanceSocketDir is the default location for forwarded sockets with a relative path in HostSocket
+	InstanceSocketDir = "sock"
+
+	// InstanceConfigDir is the default location for config files with a relative path in HostFile
+	InstanceConfigDir = "conf"
```
Comment (AkihiroSuda) [inline pkg/store/filenames/filenames.go:52]: I still think this name is confusing.

What about "copied-from-guest"

**Review by AkihiroSuda — CHANGES_REQUESTED** [review]
I still think "conf" ("config") is a confusing name

**Comment by AkihiroSuda** [comment]
Also please consider squashing comnits

**Comment by afbjorklund** [comment]
> I still think "conf" ("config") is a confusing name'

> Also please consider squashing commits

I can squash them when the new names are finalized, I was deliberately trying to scope it down from the last time...

Now it looks like a more generic functionality, but the requirement was only to do a `mkdir -p` and ssh a `sudo cat`.

I don't have a much better name, for a directory for $KUBECONFIG files.

It was supposed to be an internal concept, similar to the "sock" directory

Could call it "yaml", based on file type ?

Or maybe even "kube", to specialize it.

**Comment by jandubois** [comment]
> > I still think "conf" ("config") is a confusing name'
> 
> I don't have a much better name, for a directory for $KUBECONFIG files.

The use for `$KUBECONFIG` files is your immediate use case, but the feature is more general, and called `copyToHost`. I think @AkihiroSuda wants to quite literally call it `$INSTANCE_DIR/copied-from-guest`, which is pretty descriptive.

I admit that personally I find it too long, but I can see the value for users not yet familiar with the concept.

I too don't have a better suggestion.

> It was supposed to be an internal concept, similar to the "sock" directory

Anything documented in the `default.yaml` is not an internal concept, but a supported API.

That said, I thought `conf` was fine, but I see now that the concept is bigger than that. However, I'm not sure what else you would want to exfiltrate from the guest after provisioning, but config files.

> Could call it "yaml", based on file type ?

No, that doesn't work for me. Do you actually means to create different directories, based on the file extension of the copied file? What about files without an extension?
 
> Or maybe even "kube", to specialize it.

I think the idea is to generalize the name beyond this specific use case, not to specialize it even further.

Note that this is also just the default location for relative host names, so mostly useful for shared templates. For local instances I would probably use absolute host names to point to a convenient location instead.

**Comment by afbjorklund** [comment]
"yaml" was extrapolating from "sock" (i.e. based on file suffix)

"kube" was based on the default location, $HOME/.kube/config

If I understand the code correctly, using relative paths would also work.

```yaml
portForwards:
- guestSocket: "/var/run/docker.sock"
  hostSocket: "docker.sock"
```

```yaml
copyToHost:
- guest: "/etc/kubernetes/admin.conf"
  host: "kubeconfig.yaml"
```

It wouldn't help for printing out the location, though (probably opposite)

At least not without more variables, for the subdirectories of `{{.Dir}}`

Something like "client" perhaps, but it wouldn't make sense without "config"

Can go with `copied-from-guest`

Maybe OK, with the new wrapper.

File: pkg/store/filenames/filenames.go (hunk 4 above)
Comment (afbjorklund) [reply pkg/store/filenames/filenames.go]: Done

**Comment by jandubois** [comment]
> "yaml" was extrapolating from "sock" (i.e. based on file suffix)

But "sock" was not based on file suffix (except they accidentally match). It is based on the fact the the only thing you can forward are sockets. And the name was shortened to `sock` to make it more likely to stay under the maximum socket name length of 103 characters on macOS.

> Can go with copied-from-guest

Yeah, I can go with that if @AkihiroSuda feels strongly about not using `conf`.

> I don't really see the difference, I guess the sockets would be forwarded-from-guest

No, as explained above, there can only be sockets inside `sock`, but there can be any kind if files in `copied-from-guest`. And the name of the socket directory needs to be short. There is no such limitation on the "copied from guest" directory name.

> And think it looks rather long and ugly

I agree, but I can see the point of it being more self-explanatory.

Alternatively, maybe this could be solved by more extensive explanation in the `default.yaml` comments?

**Comment by afbjorklund** [comment]
I removed the global constants, so now it is up to each example to name the local files to whatever suits best...

The default.yaml shows the suggested `{{.Dir}}/copied-from-guest/myconfig` location, as an example host path.

**Comment by AkihiroSuda** [comment]
Thanks, but the commit does not seem updated

**Comment by jandubois** [comment]
> Thanks, but the commit does not seem updated

Seems updated to me. The current state of the PR now requires an absolute path for both the guest and the host filename, and no longer provides a default location.

It is not clear to me what you are requesting: should the default location be restored and set to `.../copied-from-guest`?

File: examples/default.yaml [hunk 5]
```diff
@@ -336,6 +336,13 @@ networks:
 #   hostPortRange: [1, 65535]
 # # Any port still not matched by a rule will not be forwarded (ignored)
 
+# Config files. Copied after provisioning scripts have been completed.
```
Comment (AkihiroSuda) [inline examples/default.yaml:339]:
```suggestion
# Copy files from the guest to the host. Copied after provisioning scripts have been completed.
```

**Review by AkihiroSuda — APPROVED** [review]
Thanks, LGTM but with a single nit

**Comment by afbjorklund** [comment]
This also opens for choosing a more suitable name for each example, matching the files that are copied.

**Comment by jandubois** [comment]
> This also opens for choosing a more suitable name for each example, matching the files that are copied.
> 
> Like `{{.Dir}}/kube/config.yaml`

This is actually a reason why I preferred to have a default location because this now opens the instance directory up to users creating arbitrarily named directories, which could conflict with files or directories Lima wan't to create in the future.

This was the reason we specified `sock/` as the default location for forwarded sockets instead of creating them directly in the instance dir.

**Comment by afbjorklund** [comment]
Actually they can copy them anywhere on the host, since we require it to be an absolute path and not just a filename...

**Comment by jandubois** [comment]
> Actually they can copy them anywhere on the host, since we require it to be an absolute path and not just a filename...

Yes, but users copy from the provided examples, and if our examples create random directories inside the instance directories, then that means that we implicitly endorse doing so.

**Comment by AkihiroSuda** [comment]
Any idea to move this forward?  (for v0.16 ?)

**Comment by jandubois** [comment]
> Any idea to move this forward? (for v0.16 ?)

Does this need anything more than the single line comment change you requested above, and that @afbjorklund gave the 👍 for, but hasn't pushed yet?

**Comment by afbjorklund** [comment]
Missed to update this, didn't know if the names were settled.

**Review by jandubois — APPROVED** [review]
Thanks, LGTM

**Pull request merged**